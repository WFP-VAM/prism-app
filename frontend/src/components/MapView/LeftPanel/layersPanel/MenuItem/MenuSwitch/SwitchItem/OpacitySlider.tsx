import {
  Box,
  Slider,
  Typography,
  WithStyles,
  createStyles,
  withStyles,
} from '@material-ui/core';
import { LayerType } from 'config/types';
import { mapSelector } from 'context/mapStateSlice/selectors';
import { opacitySelector, setOpacity } from 'context/opacityStateSlice';
import React, { ChangeEvent, memo, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';

const styles = createStyles({
  opacityText: {
    color: '#4CA1AD',
    marginBottom: '10px',
  },
  opacitySliderRoot: {
    color: '#4CA1AD',
    height: 8,
  },
  opacitySliderThumb: {
    backgroundColor: '#4CA1AD',
  },
});

interface OpacitySliderProps extends WithStyles<typeof styles> {
  activeLayerId: string;
  layerId: string;
  layerType: LayerType['type'];
}
const OpacitySlider = ({
  classes,
  activeLayerId,
  layerId,
  layerType,
}: OpacitySliderProps) => {
  const dispatch = useDispatch();
  const opacity = useSelector(opacitySelector(activeLayerId || layerId));
  const map = useSelector(mapSelector);

  const handleOnChangeSliderValue = useCallback(
    (event: ChangeEvent<{}>, newValue: number | number[]) => {
      dispatch(
        setOpacity({
          map,
          value: newValue as number,
          layerId: activeLayerId || layerId,
          layerType,
        }),
      );
    },
    [activeLayerId, dispatch, layerId, layerType, map],
  );

  return (
    <Box display="flex" justifyContent="right" alignItems="center">
      <Box pr={3}>
        <Typography
          classes={{ root: classes.opacityText }}
        >{`Opacity ${Math.round((opacity || 0) * 100)}%`}</Typography>
      </Box>
      <Box width="25%" pr={3}>
        <Slider
          value={opacity}
          step={0.01}
          min={0}
          max={1}
          aria-labelledby="left-opacity-slider"
          classes={{
            root: classes.opacitySliderRoot,
            thumb: classes.opacitySliderThumb,
          }}
          onChange={handleOnChangeSliderValue}
        />
      </Box>
    </Box>
  );
};

export default memo(withStyles(styles)(OpacitySlider));
