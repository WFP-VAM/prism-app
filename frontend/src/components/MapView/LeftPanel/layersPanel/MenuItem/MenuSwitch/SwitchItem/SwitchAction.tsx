import {
  Switch,
  WithStyles,
  createStyles,
  withStyles,
} from '@material-ui/core';
import React, { ChangeEvent, memo, useCallback } from 'react';

const styles = () =>
  createStyles({
    switch: {
      marginRight: 2,
    },
    switchTrack: {
      backgroundColor: '#E0E0E0',
    },
    switchBase: {
      color: '#E0E0E0',
      '&.Mui-checked': {
        color: '#53888F',
      },
      '&.Mui-checked + .MuiSwitch-track': {
        backgroundColor: '#B1D6DB',
      },
    },
  });

interface SwitchActionProps extends WithStyles<typeof styles> {
  someLayerAreSelected: boolean;
  toggleLayerValue: (selectedLayerId: string, checked: boolean) => void;
  activeLayerId: string;
  validatedTitle: string;
}
const SwitchAction = ({
  classes,
  someLayerAreSelected,
  toggleLayerValue,
  activeLayerId,
  validatedTitle,
}: SwitchActionProps) => {
  const handleOnChangeSwitch = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      toggleLayerValue(activeLayerId, event.target.checked);
    },
    [activeLayerId, toggleLayerValue],
  );
  return (
    <Switch
      size="small"
      className={classes.switch}
      classes={{
        switchBase: classes.switchBase,
        track: classes.switchTrack,
      }}
      checked={someLayerAreSelected}
      onChange={handleOnChangeSwitch}
      inputProps={{
        'aria-label': validatedTitle,
      }}
    />
  );
};

export default memo(withStyles(styles)(SwitchAction));
