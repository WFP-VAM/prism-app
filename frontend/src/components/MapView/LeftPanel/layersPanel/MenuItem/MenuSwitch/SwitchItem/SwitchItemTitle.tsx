import {
  MenuItem,
  Select,
  Typography,
  createStyles,
  withStyles,
} from '@material-ui/core';
import { WithStyles } from '@material-ui/styles';
import { LayerType, MenuGroupItem } from 'config/types';
import React, { memo, useCallback } from 'react';
import { useTranslation } from 'react-i18next';

const styles = () =>
  createStyles({
    title: {
      lineHeight: 1.8,
      color: 'black',
      fontWeight: 400,
    },
    titleUnchecked: {
      lineHeight: 1.8,
      color: '#828282',
      fontWeight: 400,
    },
    select: {
      '&::before': {
        border: 'none',
      },
    },
    selectItem: {
      whiteSpace: 'normal',
      fontSize: 13,
      fontWeight: 300,
      color: 'black',
      padding: 0,
      marginLeft: 5,
    },
    selectItemUnchecked: {
      whiteSpace: 'normal',
      fontSize: 13,
      fontWeight: 300,
      color: '#828282',
      padding: 0,
      marginLeft: 5,
    },
  });

const getFilteredMenuGroupItems = (menus: MenuGroupItem[], filter?: string) =>
  menus.filter(menu => (filter ? menu.id === filter : true));

interface SwitchTitleProps extends WithStyles<typeof styles> {
  layer: LayerType;
  someLayerAreSelected: boolean;
  toggleLayerValue: (selectedLayerId: string, checked: boolean) => void;
  validatedTitle: string;
  activeLayerId: string;
  setActiveLayerId: (v: string) => void;
  groupMenuFilter?: string;
  disabledMenuSelection?: boolean;
}
const SwitchItemTitle = ({
  layer,
  someLayerAreSelected,
  toggleLayerValue,
  validatedTitle,
  activeLayerId,
  setActiveLayerId,
  groupMenuFilter,
  disabledMenuSelection = false,
  classes,
}: SwitchTitleProps) => {
  const { t } = useTranslation();
  const { group } = layer;

  const handleSelect = useCallback(
    (event: React.ChangeEvent<{ value: string | unknown }>) => {
      const selectedId = event.target.value;
      setActiveLayerId(selectedId as string);
      toggleLayerValue(selectedId as string, true);
    },
    [setActiveLayerId, toggleLayerValue],
  );

  return (
    <>
      <Typography
        className={
          someLayerAreSelected ? classes.title : classes.titleUnchecked
        }
      >
        {validatedTitle}
      </Typography>
      {group && !group.activateAll && (
        <Select
          className={classes.select}
          classes={{
            root: someLayerAreSelected
              ? classes.selectItem
              : classes.selectItemUnchecked,
          }}
          value={activeLayerId}
          onChange={e => handleSelect(e)}
          disabled={disabledMenuSelection}
        >
          {getFilteredMenuGroupItems(group.layers, groupMenuFilter).map(
            menu => {
              return (
                <MenuItem key={menu.id} value={menu.id}>
                  {t(menu.label)}
                </MenuItem>
              );
            },
          )}
        </Select>
      )}
    </>
  );
};

export default memo(withStyles(styles)(SwitchItemTitle));
