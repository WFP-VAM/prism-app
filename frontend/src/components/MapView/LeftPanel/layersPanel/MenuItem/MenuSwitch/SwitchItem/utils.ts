import { Map as MapBoxMap } from 'mapbox-gl';
import { Dispatch } from 'react';
import { UrlLayerKey } from '../../../../../../../utils/url-utils';
import { LayerKey, LayerType } from '../../../../../../../config/types';
import { removeLayer } from '../../../../../../../context/mapStateSlice';
import {
  getDisplayBoundaryLayers,
  LayerDefinitions,
} from '../../../../../../../config/utils';
import {
  safeDispatchAddLayer,
  safeDispatchRemoveLayer,
} from '../../../../../../../utils/map-utils';

export function toggleRemoveLayer(
  layer: LayerType,
  _map: MapBoxMap | undefined,
  urlLayerKey: UrlLayerKey,
  dispatcher: Dispatch<any>,
  removeLayerFromUrl: Function,
) {
  removeLayerFromUrl(urlLayerKey, layer.id);
  dispatcher(removeLayer(layer));

  // For admin boundary layers with boundary property
  // we have to de-activate the unique boundary and re-activate
  // default boundaries
  if (!('boundary' in layer)) {
    return;
  }
  const boundaryId = layer.boundary || '';
  if (!Object.keys(LayerDefinitions).includes(boundaryId)) {
    return;
  }
  const displayBoundaryLayers = getDisplayBoundaryLayers();
  const uniqueBoundaryLayer = LayerDefinitions[boundaryId as LayerKey];

  if (!displayBoundaryLayers.map(l => l.id).includes(uniqueBoundaryLayer.id)) {
    safeDispatchRemoveLayer(_map, uniqueBoundaryLayer, dispatcher);
  }

  displayBoundaryLayers.forEach(l => {
    safeDispatchAddLayer(_map, l, dispatcher);
  });
}
