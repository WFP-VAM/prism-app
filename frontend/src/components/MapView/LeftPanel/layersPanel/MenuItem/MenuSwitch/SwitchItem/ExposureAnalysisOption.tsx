import { IconButton, Tooltip } from '@material-ui/core';
import { ImageAspectRatioOutlined } from '@material-ui/icons';
import { useDispatch, useSelector } from 'react-redux';
import {
  AggregationOperations,
  ExposedPopulationDefinition,
  GeometryType,
  LayerType,
  Panel,
} from 'config/types';
import { ReportsDefinitions, TableKey } from 'config/utils';
import {
  analysisResultSelector,
  clearAnalysisResult,
  ExposedPopulationDispatchParams,
  requestAndStoreExposedPopulation,
  setCurrentDataDefinition,
  setExposureLayerId,
} from 'context/analysisResultStateSlice';
import { setTabValue } from 'context/leftPanelStateSlice';
import { dateRangeSelector } from 'context/mapStateSlice/selectors';
import { useSafeTranslation } from 'i18n';
import { Extent } from 'components/MapView/Layers/raster-utils';
import { generateUniqueTableKey } from 'components/MapView/utils';

function ExposureAnalysisOption({
  layer,
  extent,
  selected,
  exposure,
}: ExposureAnalysisOptionProps) {
  const dispatch = useDispatch();
  const { t } = useSafeTranslation();
  const analysisResult = useSelector(analysisResultSelector);
  const { startDate: selectedDate } = useSelector(dateRangeSelector);

  const foundReports = Object.keys(ReportsDefinitions).filter(
    reportDefinitionKey =>
      ReportsDefinitions[reportDefinitionKey].layerId === layer.id,
  );
  if (!exposure || !foundReports.length) {
    return null;
  }

  const handleExposureAnalysis = () => {
    if (analysisResult) {
      dispatch(clearAnalysisResult());
    }

    if (!layer.id || !extent || !exposure) {
      return;
    }

    if (!selectedDate) {
      throw new Error('Date must be given to run analysis');
    }

    const hazardLayer =
      layer.type === 'wms' && layer.geometry === GeometryType.Polygon
        ? { wfsLayerId: layer.id }
        : { maskLayerId: layer.id };

    const params: ExposedPopulationDispatchParams = {
      exposure,
      date: selectedDate,
      statistic: AggregationOperations.Sum,
      extent,
      ...hazardLayer,
    };

    // Set the exposure layer id in redux so that we can have access to the reports configurations through the layer id
    dispatch(setExposureLayerId(layer.id));

    dispatch(requestAndStoreExposedPopulation(params));
    dispatch(
      setCurrentDataDefinition({
        id: generateUniqueTableKey('exposure_analysis') as TableKey,
        title: analysisResult?.getTitle(t) || '',
        table: '',
        legendText: t(analysisResult?.legendText || ''),
      }),
    );
    dispatch(setTabValue(Panel.Analysis));
  };

  return (
    <Tooltip title={t('Exposure Analysis') ?? ''}>
      <span>
        <IconButton
          id={layer.id}
          aria-label="Exposure Analysis"
          disabled={!selected}
          onClick={handleExposureAnalysis}
        >
          <ImageAspectRatioOutlined />
        </IconButton>
      </span>
    </Tooltip>
  );
}

interface ExposureAnalysisOptionProps {
  layer: LayerType;
  extent: Extent | undefined;
  selected: boolean;
  exposure: ExposedPopulationDefinition | undefined;
}

export default ExposureAnalysisOption;
