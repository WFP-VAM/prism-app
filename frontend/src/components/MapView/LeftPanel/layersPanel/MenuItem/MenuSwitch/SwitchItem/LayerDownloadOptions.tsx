import {
  Box,
  CircularProgress,
  IconButton,
  Menu,
  MenuItem,
  Tooltip,
} from '@material-ui/core';
import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { mapValues } from 'lodash';
import GetAppIcon from '@material-ui/icons/GetApp';
import {
  AdminLevelDataLayerProps,
  LayerKey,
  WMSLayerProps,
} from 'config/types';
import {
  dateRangeSelector,
  layerDataSelector,
} from 'context/mapStateSlice/selectors';
import { LayerData } from 'context/layers/layer-data';
import { downloadToFile } from 'components/MapView/utils';
import { castObjectsArrayToCsv } from 'utils/csv-utils';
import {
  downloadGeotiff,
  Extent,
} from 'components/MapView/Layers/raster-utils';
import { useSafeTranslation } from 'i18n';
import { isExposureAnalysisLoadingSelector } from 'context/analysisResultStateSlice';
import { availableDatesSelector } from 'context/serverStateSlice';
import { getRequestDate } from 'utils/server-utils';
import { LayerDefinitions, getStacBand } from 'config/utils';
import { getDateFormat } from 'utils/date-utils';
import { safeCountry } from 'config';

// TODO - return early when the layer is not selected.
function LayerDownloadOptions({
  layerId,
  extent,
  selected,
  size,
}: LayerDownloadOptionsProps) {
  const { t } = useSafeTranslation();
  const dispatch = useDispatch();
  const layer = LayerDefinitions[layerId] || Object.values(LayerDefinitions)[0];

  const [
    downloadMenuAnchorEl,
    setDownloadMenuAnchorEl,
  ] = useState<HTMLElement | null>(null);
  const [isGeotiffLoading, setIsGeotiffLoading] = useState(false);
  const isAnalysisExposureLoading = useSelector(
    isExposureAnalysisLoadingSelector,
  );

  const { startDate: selectedDate } = useSelector(dateRangeSelector);
  const serverAvailableDates = useSelector(availableDatesSelector);
  const layerAvailableDates = serverAvailableDates[layer.id];
  const queryDate = selected
    ? getRequestDate(layerAvailableDates, selectedDate)
    : undefined;

  const adminLevelLayerData = useSelector(
    layerDataSelector(layer.id, queryDate),
  ) as LayerData<AdminLevelDataLayerProps>;

  const handleDownloadMenuClose = () => {
    setDownloadMenuAnchorEl(null);
  };

  const handleDownloadMenuOpen = (event: React.MouseEvent<HTMLElement>) => {
    setDownloadMenuAnchorEl(event.currentTarget);
  };

  const getFilename = (): string => {
    const safeTitle = layer.title ?? layer.id;
    if (selectedDate && (layer as AdminLevelDataLayerProps).dates) {
      const dateString = getDateFormat(selectedDate, 'snake');
      return `${safeTitle}_${dateString}`;
    }
    return safeTitle;
  };

  const handleDownloadGeoJson = (): void => {
    if (!adminLevelLayerData) {
      console.warn(`No layer data available for ${layer.id}`);
    }
    downloadToFile(
      {
        content: JSON.stringify(adminLevelLayerData?.data.features),
        isUrl: false,
      },
      getFilename(),
      'application/json',
    );
    handleDownloadMenuClose();
  };

  const handleDownloadCsv = (): void => {
    if (!adminLevelLayerData) {
      console.warn(`No layer data available for ${layer.id}`);
    }
    const translatedColumnsNames = mapValues(
      adminLevelLayerData?.data.layerData[0],
      (v, k) => (k === 'value' ? t(adminLevelLayerData.layer.id) : t(k)),
    );
    downloadToFile(
      {
        content: castObjectsArrayToCsv(
          adminLevelLayerData?.data.layerData,
          translatedColumnsNames,
          ';',
        ),
        isUrl: false,
      },
      getFilename(),
      'text/csv',
    );
    handleDownloadMenuClose();
  };

  const handleDownloadGeoTiff = (): void => {
    const { serverLayerName, additionalQueryParams } = layer as WMSLayerProps;
    const band = getStacBand(additionalQueryParams);
    const dateString = getDateFormat(selectedDate, 'default') as string;

    setIsGeotiffLoading(true);
    downloadGeotiff(
      serverLayerName,
      band,
      extent,
      dateString,
      `${safeCountry}_${layerId}_${dateString}.tif`,
      dispatch,
      () => setIsGeotiffLoading(false),
    );
    handleDownloadMenuClose();
  };

  const shouldShowDownloadButton =
    layer.type === 'admin_level_data' ||
    (layer.type === 'wms' &&
      layer.baseUrl.includes('api.earthobservation.vam.wfp.org/ows'));

  return (
    <>
      {shouldShowDownloadButton && (
        <Tooltip title="Download">
          <span>
            <IconButton
              disabled={!selected || isGeotiffLoading}
              onClick={handleDownloadMenuOpen}
              size={size || 'medium'}
            >
              <GetAppIcon fontSize={size || 'medium'} />
            </IconButton>
          </span>
        </Tooltip>
      )}
      {isGeotiffLoading ||
        (isAnalysisExposureLoading && (
          <Box display="flex" alignItems="center">
            <CircularProgress size="20px" />
          </Box>
        ))}
      <Menu
        id="download-menu"
        anchorEl={downloadMenuAnchorEl}
        keepMounted
        open={Boolean(downloadMenuAnchorEl)}
        onClose={handleDownloadMenuClose}
      >
        {layer.type === 'admin_level_data' && [
          <MenuItem key="download-as-csv" onClick={handleDownloadCsv}>
            {t('Download as CSV')}
          </MenuItem>,
          <MenuItem key="download-as-geojson" onClick={handleDownloadGeoJson}>
            {t('Download as GeoJSON')}
          </MenuItem>,
        ]}
        {layer.type === 'wms' &&
          layer.baseUrl.includes('api.earthobservation.vam.wfp.org/ows') && (
            <MenuItem key="download-as-geotiff" onClick={handleDownloadGeoTiff}>
              {t('Download as GeoTIFF')}
            </MenuItem>
          )}
      </Menu>
    </>
  );
}

interface LayerDownloadOptionsProps {
  layerId: LayerKey;
  extent: Extent | undefined;
  selected: boolean;
  size?: 'small' | undefined;
}

export default LayerDownloadOptions;
