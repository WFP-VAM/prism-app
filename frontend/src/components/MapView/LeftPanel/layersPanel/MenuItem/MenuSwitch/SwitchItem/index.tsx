import {
  Box,
  createStyles,
  IconButton,
  Tooltip,
  WithStyles,
  withStyles,
} from '@material-ui/core';
import OpacityIcon from '@material-ui/icons/Opacity';
import React, { memo, useCallback, useEffect, useMemo, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { LayerKey, LayerType } from 'config/types';
import { LayerDefinitions } from 'config/utils';
import { clearDataset } from 'context/datasetStateSlice';
import { layersSelector, mapSelector } from 'context/mapStateSlice/selectors';
import { useSafeTranslation } from 'i18n';
import { refreshBoundaries } from 'utils/map-utils';
import { getUrlKey, useUrlHistory } from 'utils/url-utils';
import { Extent } from 'components/MapView/Layers/raster-utils';
import { availableDatesSelector } from 'context/serverStateSlice';
import { checkLayerAvailableDatesAndContinueOrRemove } from 'components/MapView/utils';
import { LocalError } from 'utils/error-utils';
import { opacitySelector, setOpacity } from 'context/opacityStateSlice';
import { toggleRemoveLayer } from './utils';
import LayerDownloadOptions from './LayerDownloadOptions';
import ExposureAnalysisOption from './ExposureAnalysisOption';
import SwitchTitle from './SwitchItemTitle';
import SwitchAction from './SwitchAction';
import OpacitySlider from './OpacitySlider';

const SwitchItem = memo(
  ({
    classes,
    layer,
    extent,
    groupMenuFilter,
    disabledMenuSelection,
  }: SwitchItemProps) => {
    const {
      id: layerId,
      title: layerTitle,
      opacity: initialOpacity,
      type: layerType,
      group,
    } = layer;
    const { t } = useSafeTranslation();
    const selectedLayers = useSelector(layersSelector);
    const serverAvailableDates = useSelector(availableDatesSelector);
    const map = useSelector(mapSelector);
    const [isOpacitySelected, setIsOpacitySelected] = useState(false);
    const dispatch = useDispatch();
    const opacity = useSelector(opacitySelector(layerId));
    const {
      updateHistory,
      appendLayerToUrl,
      removeLayerFromUrl,
    } = useUrlHistory();

    useEffect(() => {
      setIsOpacitySelected(false);
    }, [dispatch, initialOpacity, layerId, layerType, map]);

    useEffect(() => {
      if (opacity !== undefined) {
        return;
      }
      dispatch(
        setOpacity({
          map,
          value: initialOpacity || 0,
          layerId,
          layerType,
        }),
      );
    }, [dispatch, initialOpacity, layerId, layerType, map, opacity]);

    const someLayerAreSelected = useMemo(() => {
      return selectedLayers.some(
        ({ id: testId }) =>
          testId === layerId ||
          (group && group.layers.some(l => l.id === testId)),
      );
    }, [group, layerId, selectedLayers]);

    const selectedActiveLayer = useMemo(
      () =>
        someLayerAreSelected
          ? selectedLayers.filter(
              sl =>
                (group?.activateAll &&
                  group?.layers.find(l => l.id === sl.id && l.main)) ||
                (!group?.activateAll &&
                  group?.layers.map(l => l.id).includes(sl.id)),
            )
          : [],
      [group, someLayerAreSelected, selectedLayers],
    );

    const initialActiveLayerId = useMemo(() => {
      return selectedActiveLayer.length > 0
        ? selectedActiveLayer[0].id
        : layer.id;
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [layer.id]);

    const [activeLayerId, setActiveLayerId] = useState(
      initialActiveLayerId || (group?.layers?.find(l => l.main)?.id as string),
    );

    useEffect(() => {
      setActiveLayerId(
        initialActiveLayerId ||
          (group?.layers?.find(l => l.main)?.id as string),
      );
    }, [group, initialActiveLayerId]);

    const exposure = useMemo(() => {
      return (layer.type === 'wms' && layer.exposure) || undefined;
    }, [layer.exposure, layer.type]);

    const validatedTitle = useMemo(() => {
      return t(group?.groupTitle || layerTitle || '');
    }, [group, layerTitle, t]);

    const toggleLayerValue = useCallback(
      (selectedLayerId: string, checked: boolean) => {
        // reset opacity selected
        setIsOpacitySelected(false);
        // clear previous table dataset loaded first
        // to close the dataseries and thus close chart
        dispatch(clearDataset());

        const selectedLayer = group
          ? LayerDefinitions[selectedLayerId as LayerKey]
          : layer;

        const urlLayerKey = getUrlKey(selectedLayer);

        if (!checked) {
          toggleRemoveLayer(
            selectedLayer,
            map,
            urlLayerKey,
            dispatch,
            removeLayerFromUrl,
          );
          return;
        }
        try {
          checkLayerAvailableDatesAndContinueOrRemove(
            layer,
            serverAvailableDates,
            removeLayerFromUrl,
            dispatch,
          );
        } catch (error) {
          console.error((error as LocalError).getErrorMessage());
          return;
        }
        const updatedUrl = appendLayerToUrl(
          urlLayerKey,
          selectedLayers,
          selectedLayer,
        );
        updateHistory(urlLayerKey, updatedUrl);
        if (
          'boundary' in selectedLayer ||
          selectedLayer.type !== 'admin_level_data'
        ) {
          return;
        }
        refreshBoundaries(map, dispatch);
      },
      [
        appendLayerToUrl,
        dispatch,
        group,
        layer,
        map,
        removeLayerFromUrl,
        selectedLayers,
        serverAvailableDates,
        updateHistory,
      ],
    );

    return (
      <Box display="flex" flexDirection="column" maxWidth="100%">
        <Box key={layerId} display="flex" alignItems="center" m={2}>
          <SwitchAction
            activeLayerId={activeLayerId}
            someLayerAreSelected={someLayerAreSelected}
            toggleLayerValue={toggleLayerValue}
            validatedTitle={validatedTitle}
          />
          <SwitchTitle
            layer={layer}
            activeLayerId={activeLayerId}
            someLayerAreSelected={someLayerAreSelected}
            toggleLayerValue={toggleLayerValue}
            setActiveLayerId={setActiveLayerId}
            validatedTitle={validatedTitle}
            groupMenuFilter={groupMenuFilter}
            disabledMenuSelection={disabledMenuSelection}
          />
          <Tooltip title="Opacity">
            <span style={{ marginLeft: 'auto' }}>
              <IconButton
                disabled={!someLayerAreSelected}
                classes={{
                  root: isOpacitySelected
                    ? classes.opacityRootSelected
                    : classes.opacityRoot,
                }}
                onClick={() =>
                  setIsOpacitySelected(opacitySelected => !opacitySelected)
                }
              >
                <OpacityIcon />
              </IconButton>
            </span>
          </Tooltip>
          <ExposureAnalysisOption
            exposure={exposure}
            extent={extent}
            layer={layer}
            selected={someLayerAreSelected}
          />
          <LayerDownloadOptions
            layerId={activeLayerId}
            extent={extent}
            selected={someLayerAreSelected}
          />
        </Box>
        {someLayerAreSelected && isOpacitySelected && (
          <OpacitySlider
            activeLayerId={activeLayerId}
            layerId={layerId}
            layerType={layerType}
          />
        )}
      </Box>
    );
  },
);

const styles = () =>
  createStyles({
    switch: {
      marginRight: 2,
    },
    switchTrack: {
      backgroundColor: '#E0E0E0',
    },
    switchBase: {
      color: '#E0E0E0',
      '&.Mui-checked': {
        color: '#53888F',
      },
      '&.Mui-checked + .MuiSwitch-track': {
        backgroundColor: '#B1D6DB',
      },
    },
    opacityRoot: {
      color: '#828282',
      marginLeft: 'auto',
    },
    opacityRootSelected: {
      backgroundColor: '#4CA1AD',
      color: '#F2F2F2',
      marginLeft: 'auto',
      '&:hover': {
        color: '#4CA1AD',
      },
    },
  });

export interface SwitchItemProps extends WithStyles<typeof styles> {
  layer: LayerType;
  extent?: Extent;
  groupMenuFilter?: string;
  disabledMenuSelection?: boolean;
}

export default withStyles(styles)(SwitchItem);
