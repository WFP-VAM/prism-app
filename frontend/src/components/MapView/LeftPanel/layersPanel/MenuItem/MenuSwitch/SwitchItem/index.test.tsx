import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import { store } from 'context/store';
import { LayerType } from 'config/types';
import SwitchItem from '.';

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    replace: jest.fn(),
    location: {
      search: '',
    },
  }),
}));

const props = {
  layer: {} as LayerType,
};

test('renders as expected', () => {
  const { container } = render(
    <Provider store={store}>
      <SwitchItem {...props} />
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
