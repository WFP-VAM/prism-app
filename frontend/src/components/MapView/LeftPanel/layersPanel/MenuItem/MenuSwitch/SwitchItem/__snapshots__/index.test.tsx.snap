// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-6"
    style="display: flex; flex-direction: column; max-width: 100%;"
  >
    <div
      class="MuiBox-root MuiBox-root-7"
      style="display: flex; align-items: center;"
    >
      <mock-switch
        checked="false"
        classes="[object Object]"
        classname="makeStyles-switch-8"
        color="primary"
        inputprops="[object Object]"
      />
      <mock-typography
        classname="makeStyles-titleUnchecked-13"
      />
      <mock-tooltip
        title="Opacity"
      >
        <span
          style="margin-left: auto;"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-opacityRoot-4 Mui-disabled Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17.66 8L12 2.35 6.34 8C4.78 9.56 4 11.64 4 13.64s.78 4.11 2.34 5.67 3.61 2.35 5.66 2.35 4.1-.79 5.66-2.35S20 15.64 20 13.64 19.22 9.56 17.66 8zM6 14c.01-2 .62-3.27 1.76-4.4L12 5.27l4.24 4.38C17.38 10.77 17.99 12 18 14H6z"
                />
              </svg>
            </span>
          </button>
        </span>
      </mock-tooltip>
      <mock-menu
        id="download-menu"
        keepmounted="true"
        open="false"
      />
    </div>
  </div>
</div>
`;
